runtime/lib/ttnn/runtime.cpp

```diff
@@ -29,39 +29,69 @@ static ::tt::target::Dim2d toFlatbuffer(CoreCoord coreCoord) {
   return ::tt::target::Dim2d(coreCoord.y, coreCoord.x);
 }
 
-std::pair<SystemDesc, DeviceIds> getCurrentSystemDesc() {
-  auto &device = ::ttnn::open_device(0);
-  std::vector<int> chipIds = {
-      device.id(),
-  };
+static std::pair<int, int> getConnectedChips(const ::ttnn::Device &device) {
+  std::unordered_set<chip_id_t> connectedChips =
+      device.get_ethernet_connected_device_ids();
+  std::vector<chip_id_t> connectedChipsOrdered(connectedChips.begin(),
+                                               connectedChips.end());
+  std::sort(connectedChipsOrdered.begin(), connectedChipsOrdered.end());
+  std::vector<int> connectedChipsWithDefault = {-1, -1};
+  for (size_t i = 0; i < connectedChipsOrdered.size(); i++) {
+    connectedChipsWithDefault[i] = connectedChipsOrdered[i];
+  }
+  return std::make_pair(connectedChipsWithDefault[0],
+                        connectedChipsWithDefault[1]);
+}
+
+static std::unique_ptr<std::pair<SystemDesc, DeviceIds>>
+getCurrentSystemDescImpl(std::vector<::ttnn::Device *> &mmio_devices,
+                         std::vector<::ttnn::Device *> &remote_devices) {
+  size_t numDevices = ::tt::tt_metal::GetNumAvailableDevices();
+  std::vector<int> chipIds;
+  std::vector<::flatbuffers::Offset<tt::target::ChipDesc>> chipDescs;
+  std::vector<uint32_t> chipDescIndices;
+  std::vector<::tt::target::ChipCapability> chipCapabilities;
+  std::vector<::tt::target::ChipCoord> chipCoords;
+  std::vector<::tt::target::ChipChannel> chipChannels;
   ::flatbuffers::FlatBufferBuilder fbb;
+
+  for (size_t deviceId = 0; deviceId < numDevices; deviceId++) {
+    ::ttnn::Device &device = ::ttnn::open_device(deviceId);
+    chipIds.push_back(device.id());
+    // Construct chip descriptor
+    ::tt::target::Dim2d deviceGrid = toFlatbuffer(device.logical_grid_size());
+    chipDescs.push_back(::tt::target::CreateChipDesc(
+        fbb, toFlatbuffer(device.arch()), &deviceGrid,
+        device.l1_size_per_core(), device.num_dram_channels(),
+        device.dram_size_per_channel(), L1_ALIGNMENT, PCIE_ALIGNMENT,
+        DRAM_ALIGNMENT));
+    chipDescIndices.push_back(deviceId);
+    // Derive chip capability
+    ::tt::target::ChipCapability chipCapability =
+        ::tt::target::ChipCapability::NONE;
+    if (device.is_mmio_capable()) {
+      chipCapability = chipCapability | ::tt::target::ChipCapability::PCIE |
+                       ::tt::target::ChipCapability::HostMMIO;
+      mmio_devices.push_back(&device);
+    } else {
+      remote_devices.push_back(&device);
+    }
+    chipCapabilities.push_back(chipCapability);
+    // Extract chip location
+    int x, y, rack, shelf;
+    std::tie(x, y, rack, shelf) = device.get_chip_location();
+    chipCoords.emplace_back(rack, shelf, y, x);
+    // Extract chip connected channels
+    std::pair<int, int> connectedChips = getConnectedChips(device);
```

Comment:
Can't there be more than 1 connect chips?  Seems like a remote chip in the middle of a galaxy could have up to 4 connections.

Reply:
Hmmmm that's a good point, I'll update the chipChannels schema to be an array of ints then instead of having only 2 endpoints hardcoded

Reply:
The intention for `chipChannels` was to be just a flat list of connections, not tied to the other arrays, but maybe this is more cumbersome to query.  If you think of it like a graph you have a list of nodes `chipIds` and a list of edges `chipChannels` where each edge is just a pair of node ids (but could be extended with new attributes in the future, like number of streams).

Reply:
@nsmithtt Ahhh I see, I misinterpreted what chipChannels was intended to do. I pushed an update that now constructs a flat list of connections - also ported to a metal device mesh API that handles opening/closing devices safely. 

Updated system descriptor looks like:

`{'version': {'major': 0, 'minor': 0, 'patch': 40}, 'ttmlir_git_hash': '1fdd4a0970b17925db82c74c163834ccb24c737f', 'product_identifier': 'unknown', 'system_desc': {'chip_descs': [{'arch': 'Wormhole_b0', 'grid_size': {'y': 8, 'x': 8}, 'l1_size': 1499136, 'num_dram_channels': 12, 'dram_channel_size': 1073741824, 'noc_l1_address_align_bytes': 16, 'pcie_address_align_bytes': 32, 'noc_dram_address_align_bytes': 32}, {'arch': 'Wormhole_b0', 'grid_size': {'y': 8, 'x': 8}, 'l1_size': 1499136, 'num_dram_channels': 12, 'dram_channel_size': 1073741824, 'noc_l1_address_align_bytes': 16, 'pcie_address_align_bytes': 32, 'noc_dram_address_align_bytes': 32}], 'chip_desc_indices': [0, 1], 'chip_capabilities': [3, 0], 'chip_coords': [{'rack': 0, 'shelf': 0, 'y': 0, 'x': 0}, {'rack': 0, 'shelf': 0, 'y': 0, 'x': 1}], 'chip_channels': [{'endpoint0': 0, 'endpoint1': 1}]}}`